### PR TITLE
[SYCLomatic][Helper Functions] Add an interface for customers to filter out devices in dev_mgr by names

### PR DESCRIPTION
**Pull request by ziranzha**

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 1]
```diff
@@ -788,6 +811,12 @@ select_device(const DeviceSelector &selector = sycl::gpu_selector_v) {
   dev_mgr::instance().select_device(selector);
 }
 
+/// Filter out device that doesn't contain one of the subnames.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]:
```suggestion
/// Filter out devices; only keep the device with the name specified.
/// \p dev_subnames: subname of device, which will be kept.
```

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 2]
```diff
@@ -788,6 +811,12 @@ select_device(const DeviceSelector &selector = sycl::gpu_selector_v) {
...
+/// May break device id mapping and change current device.
+static inline void filter_device(const std::vector<std::string> &subnames) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: ### 
```suggestion
static inline void filter_device(const std::vector<std::string> &dev_subnames) {
```

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 3]
```diff
@@ -659,19 +658,43 @@ class dev_mgr {
...
     }
-    return id;
+    throw std::runtime_error("Cannot find the device[" +
+                             dev.get_info<sycl::info::device::name>() +
+                             "] in current device list!")
+  }
+
+/// Filter out device that doesn't contain one of the subnames.
+/// May break device id mapping and change current device.
+  void filter(const std::vector<std::string> &subnames) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: logic error, subnames is not used.

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 4]
```diff
@@ -628,9 +628,8 @@ static inline unsigned int get_tid() {
 class dev_mgr {
 public:
   device_ext &current_device() {
-    unsigned int dev_id=current_device_id();
-    check_id(dev_id);
-    return *_devs[dev_id];
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: get_device will hold the lock, so L631 is not required.

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 5]
```diff
@@ -659,19 +658,51 @@ class dev_mgr {
...
     }
-    return id;
+    throw std::runtime_error(
+        "The device[" + dev.get_info<sycl::info::device::name>() +
+        "] is filtered out by dev_mgr::filter/filter_device in current device "
+        "list!");
+  }
+
+  /// Filter out devices; only keep the device with the name specified.
+  /// May break device id mapping and change current device.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: For L680, please be more specific on what is: id mapping.  suggest provide a func to print the dev-id mapping.

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 6]
```diff
@@ -659,19 +658,51 @@ class dev_mgr {
...
+        iter = _devs.erase(iter);
+    }
+    _cpu_device = -1;
+    for (unsigned i = 0; i < _devs.size(); ++i) {
+      if (_devs[i]->is_cpu()) {
+        _cpu_device = i;
+        break;
+      }
+    }
+    _thread2dev_map.clear();
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp:705]: suggest to emit exception if dev list is empty after filter().

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 7]
```diff
@@ -788,6 +819,12 @@ select_device(const DeviceSelector &selector = sycl::gpu_selector_v) {
   dev_mgr::instance().select_device(selector);
 }
 
+/// Filter out devices; only keep the device with the name specified.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]: Please suggest the usage of the static function.

File: clang/runtime/dpct-rt/include/dpct/device.hpp [hunk 8]
```diff
@@ -653,25 +651,71 @@ class dev_mgr {
...
+    }
+    _cpu_device = -1;
+    for (unsigned i = 0; i < _devs.size(); ++i) {
+      if (_devs[i]->is_cpu()) {
+        _cpu_device = i;
+        break;
+      }
+    }
+    _thread2dev_map.clear();
+#ifdef DPCT_VERBOSE
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/device.hpp]:
```suggestion
#ifdef DPCT_HELPER_VERBOSE
```

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**